### PR TITLE
Backport/colton delivery ai performance

### PR DESCRIPTION
**Pull request by Raycoms**
Like it says in the title @cltnschlosser please

Review please

### CODE REVIEW

**Review by xavierh — APPROVED** [review]

**Comment by Raycoms** [comment]
I think I have adapt some changes here.
Things got mixed up with the newest prs.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIMeleeGuard.java [hunk 1]
```diff
@@ -113,7 +113,7 @@ protected AIState searchTarget()
         {
             return AIState.GUARD_SEARCH_TARGET;
         }
-        InventoryFunctions.matchFirstInProvider(worker, stack -> !ItemStackUtils.isEmpty(stack) && ItemStackUtils.doesItemServeAsWeapon(stack), worker::setHeldItem);
+        InventoryFunctions.matchFirstInProviderWithSimpleAction(worker, stack -> !ItemStackUtils.isEmpty(stack) && ItemStackUtils.doesItemServeAsWeapon(stack), worker::setHeldItem);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIMeleeGuard.java:116]: MAJOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

File: src/main/java/com/minecolonies/coremod/util/InventoryFunctions.java [hunk 2]
```diff
@@ -56,9 +45,8 @@ public static boolean matchFirstInProvider(
      * @param tester   the function to use for testing slots
      * @return true if it found a stack
      */
-    private static boolean matchFirstInProvider(
-                                                 final ICapabilityProvider provider, @NotNull final Function<ICapabilityProvider, Function<Integer,
-                                                                                                                                            Predicate<ItemStack>>> tester)
+    private static boolean matchFirstInProvider(final ICapabilityProvider provider,
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/InventoryFunctions.java:48]: MAJOR Remove this unused private "matchFirstInProvider" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java [hunk 3]
```diff
@@ -502,9 +496,8 @@ else if (itemsToDeliver.isEmpty())
      * Gathers only one stack of the item.
      *
      * @param buildingToDeliver building to deliver to.
-     * @return true if continue, false if not succuesful
      */
-    private boolean gatherItems(@NotNull final AbstractBuilding buildingToDeliver, @NotNull final BlockPos position)
+    private AIState gatherItems(@NotNull final AbstractBuilding buildingToDeliver, @NotNull final BlockPos position)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java:500]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
MAJOR Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 4]
```diff
@@ -192,7 +201,7 @@ protected void onException(final RuntimeException e)
             // fix for printing the actual exception
             e.printStackTrace();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:202]: MINOR Use a logger to log this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 5]
```diff
@@ -192,7 +201,7 @@ protected void onException(final RuntimeException e)
...
         }
-        catch (RuntimeException exp)
+        catch (final RuntimeException exp)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:204]: MAJOR Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 6]
```diff
@@ -192,7 +201,7 @@ protected void onException(final RuntimeException e)
...
         {
             Log.getLogger().error("Welp reporting crashed:");
             exp.printStackTrace();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:207]: MINOR Use a logger to log this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1148)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 58 issues
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 41 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CitizenData.java#L23](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L23): class "CitizenData" has 43 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIBasic.java#L39](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java#L39): class "AbstractEntityAIBasic" has 55 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [AbstractEntityAIBasic.java#L532](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java#L532): Remove this unused private "requestWithoutSpam" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [AbstractEntityAIBasic.java#L983](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java#L983): Provide multiple methods instead of using "useItemDamage" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [AbstractEntityAIBasic.java#L1032](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java#L1032): Provide multiple methods instead of using "useItemDamage" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [TileEntityWareHouse.java#L162](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java#L162): Refactor this method to reduce its Cognitive Complexity from 26 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [InventoryUtils.java#L24](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/coremod/util/InventoryUtils.java#L24): class "InventoryUtils" has 75 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MINOR][MINOR] [blockout](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/blockout): Add a 'package-info.java' file to document the 'blockout' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [controls](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/blockout/controls): Add a 'package-info.java' file to document the 'controls' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [views](https://github.com/Minecolonies/minecolonies/blob/55eeca7213b907afaf63d858790ab2782deacf12/src/main/java/com/minecolonies/blockout/views): Add a 'package-info.java' file to document the 'views' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Raycoms** [comment]
@OrionDevelopment can we?

**Review by marchermans — APPROVED** [review]
Yeah looks good.

**Pull request merged**